source/common/stats/stats_impl.h

```diff
@@ -333,27 +342,51 @@ class GaugeImpl : public Gauge, public MetricImpl {
   // Stats::Gauge
   virtual void add(uint64_t amount) override {
     data_.value_ += amount;
-    data_.flags_ |= RawStatData::Flags::Used;
+    data_.flags_ |= Flags::Used;
   }
   virtual void dec() override { sub(1); }
   virtual void inc() override { add(1); }
   virtual void set(uint64_t value) override {
     data_.value_ = value;
-    data_.flags_ |= RawStatData::Flags::Used;
+    data_.flags_ |= Flags::Used;
   }
   virtual void sub(uint64_t amount) override {
     ASSERT(data_.value_ >= amount);
     ASSERT(used());
     data_.value_ -= amount;
   }
   virtual uint64_t value() const override { return data_.value_; }
-  bool used() const override { return data_.flags_ & RawStatData::Flags::Used; }
+  bool used() const override { return data_.flags_ & Flags::Used; }
 
 private:
   RawStatData& data_;
   RawStatDataAllocator& alloc_;
 };
 
+/**
+ * Implementation of HistogramStatistics for circllhist.
+ */
+class HistogramStatisticsImpl : public HistogramStatistics {
+public:
+  HistogramStatisticsImpl() {
+    for (size_t i = 0; i < ARRAY_SIZE(quantiles_out_); i++) {
+      quantiles_out_[i] = 0;
+    }
+  }
+
+  HistogramStatisticsImpl(histogram_t* histogram_ptr) {
+    hist_approx_quantile(histogram_ptr, const_cast<double*>(quantiles_in_),
+                         ARRAY_SIZE(quantiles_in_), quantiles_out_);
+  }
+
+  std::string summary() const override {
+    return fmt::format("P0: {} , P25: {}, P50: {}, P75: {}, P90: {}, P95: {}, P99: {}, P100: {}",
```

Comment:
WDYT of constructing this by looping over the quantiles_in_ and quantiles_out_ vectors, so they'll never get out of sink. In fact it looks like they are out of sink now as you have a .999 in the quantiles_in_ initializer above; only 8 values in your format string and 9 args below.

Easy to miss.  Let's avoid that by constructing this from the source of truth.

Reply:
+1